File: solutions/problem_1928_minimum_cost_to_reach_destination_in_time/solution.py
"""
Problem 1928: Minimum Cost to Reach Destination in Time
=====================================================
Difficulty: Hard
Tags: Array, Dynamic Programming, Graph Theory
Pattern: Graph Algorithm

Time Complexity:  O(V + E)
Space Complexity: O(V + E)
"""

class Solution:
    def minCost(self, maxTime: int, edges: List[List[int]], passingFees: List[int]) -> int:
        # Graph traversal - O(V+E) time
        from collections import defaultdict
        graph = defaultdict(list)
        # Build adjacency list from input
        n = len(maxTime) if isinstance(maxTime, list) else maxTime
        visited = [False] * n
        result = 0
        
        def dfs(node):
            visited[node] = True
            for neighbor in graph[node]:
                if not visited[neighbor]:
                    dfs(neighbor)
        
        for i in range(n):
            if not visited[i]:
                dfs(i)
                result += 1
        return result

File: solutions/problem_1928_minimum_cost_to_reach_destination_in_time/solution.cpp
/*
 * Problem 1928: Minimum Cost to Reach Destination in Time
 * =====================================================
 * Difficulty: Hard
 * Tags: Array, Dynamic Programming, Graph Theory
 * Pattern: Graph Algorithm
 *
 * Time Complexity:  O(V + E)
 * Space Complexity: O(V + E)
 */

#include <functional>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    int minCost(int maxTime, vector<vector<int>>& edges, vector<int>& passingFees) {
        // Graph traversal - O(V+E) time
        int n = maxTime.size();
        vector<vector<int>> graph(n);
        vector<bool> visited(n, false);
        int result = 0;
        function<void(int)> dfs = [&](int node) {
            visited[node] = true;
            for (int neighbor : graph[node]) {
                if (!visited[neighbor]) dfs(neighbor);
            }
        };
        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                dfs(i);
                result++;
            }
        }
        return result;
    }
};
